src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/PkcsPalOpenSsl.cs

```diff
@@ -149,8 +149,50 @@ public sealed override Exception CreateDecryptTwiceException()
         /// </summary>
         public sealed override byte[] GetSubjectKeyIdentifier(X509Certificate2 certificate)
         {
-            // TODO(3334): Use certificate.PublicKey and the Subject properties with the encoder. 
-            throw new NotImplementedException();
+            X509ExtensionCollection certExtensions = certificate.Extensions;
+            foreach (X509Extension ext in certExtensions)
+            {
+                if (ext != null && ext.Oid != null &&
+                    StringComparer.Ordinal.Equals(Oids.SubjectKeyIdentifier, ext.Oid.Value))
+                {
+                    return DerSequenceReader.CreateForPayload(ext.RawData).ReadOctetString();
+                }
+            }
+
+            // If we've reached this point, it means the certificate doesn't explicitly have a SubjectKeyIdentifier
+            // extension, so one must be generated using SHA1
+
+            // As stated in OpenSslCertificateFinder:
+            // The Desktop/Windows version of this method use CertGetCertificateContextProperty
+            // with a property ID of CERT_KEY_IDENTIFIER_PROP_ID.
+            //
+            // MSDN says that when there's no extension, this method takes the SHA-1 of the
+            // SubjectPublicKeyInfo block, and returns that.
+            //
+            // https://msdn.microsoft.com/en-us/library/windows/desktop/aa376079%28v=vs.85%29.aspx
+
+            using (HashAlgorithm hash = SHA1.Create())
+            {
+                byte[] publicKeyInfo = EncodePublicKeyInfo(certificate);
+                return hash.ComputeHash(publicKeyInfo);
+            }
+        }
+
+        private byte[] EncodePublicKeyInfo(X509Certificate2 certificate)
+        {
+            // SubjectPublicKeyInfo ::= SEQUENCE {
+            //      algorithm AlgorithmIdentifier,
+            //      subjectPublicKey BIT STRING }
+
+            PublicKey pubKey = certificate.PublicKey;
+
+            byte[][] oid = DerEncoder.SegmentedEncodeOid(pubKey.Oid);
+            byte[][] algoParams = DerSequenceReader.SplitValue(pubKey.EncodedParameters.RawData);
+            byte[][] algoIdentifier = DerEncoder.ConstructSegmentedSequence(oid, algoParams);
+
+            byte[][] publicKey = DerSequenceReader.SplitValue(pubKey.EncodedKeyValue.RawData);
+
+            return DerEncoder.ConstructSequence(algoIdentifier, publicKey);
```

Comment:
While I commend you for being able to solve this in managed code; it's certainly a lot less memory copying if we just made a function in the native library which takes in an X509\* (cert.Handle would apply) and fills in a byte array.

Or you could achieve the same thing by using cert.Handle and X509Duplicate (instead of certPal.SafeHandle), and copying the code from OpenSslCertificateFinder.

But
```C++
extern "C" int32_t CryptoNative_X509GetImplicitSubjectKeyIdentifier(X509* x509, int8_t* buffer, int32_t* bufferSize)
{
    ...
    if (*bufferSize != MD_SHA1_SIZE)
    {
        return someError;
    }
    ...
}
```

seems better.